src/convert.rs

```diff
@@ -46,6 +46,27 @@ pub trait FromWasmAbi: WasmDescribe {
     unsafe fn from_abi(js: Self::Abi, extra: &mut Stack) -> Self;
 }
 
+/// A trait for anything that can be recovered by-value from the wasm ABI
+/// but whose conversion might error, for example a JS string being converted
+/// to a Rust enum.
+pub trait TryFromWasmAbi: WasmDescribe + Sized {
```

Comment:
I'm a little wary of adding a trait like this, I'm afraid it'll be difficult to work with in general. The conversion traits are used in various locations throughout imports/exports/closures/etc. This also necessitates `FromWasmAbi` for `Result` which seems unfortunate in the sense that you can't do:

```rust
#[wasm_bindgen]
pub fn foo(a: CustomEnumType) {
    // ...
}
```

but are instead forced to do:

```rust
#[wasm_bindgen]
pub fn foo(a: Result<CustomEnumType, String>) {
    // ...
}
```

I might be leaning a bit towards dealing with nonexhaustive enums due to this interaction of how to add implicit conversions? I would also be ok, though, having a fallible constructor in Rust but panicking in the conversions at the boundary (as it seems like that's what may happen most of the time anyway?)

Reply:
Non-exhaustive enums are fine by me, the only awkward thing with them is the potential panic if the Rust code gets an invalid enum then tries to return it back to JS, which I can't think of a decent way around.